modules/Multimode-Networks/src/main/java/cz/cvut/fit/gephi/multimode/LongTaskTransformation.java

```diff
@@ -0,0 +1,193 @@
+        List<Node> secondHorizontal = new ArrayList<Node>();
+        List<Node> secondVertical = new ArrayList<Node>();
+        for (Node n : nodes) {
+            String nodeValue;
+            Object val = n.getAttribute(attributeColumn);
+            if (val != null) {
+                nodeValue = val.toString();
+            } else {
+                nodeValue = "null";
+            }
+            // matrix axis
+            if (nodeValue.equals(inDimension)) {
+                firstVertical.add(n);
+            }
+            if (nodeValue.equals(commonDimension)) {
+                firstHorizontal.add(n);
+                secondVertical.add(n);
+            }
+            if (nodeValue.equals(outDimension)) {
+                secondHorizontal.add(n);
+            }
+        }
+        
+        if (cancelled) {
+            return;
+        }
+        Progress.progress(progressTicket);
+
+        // first matrix
+        Matrix firstMatrix = new Matrix(firstVertical.size(), firstHorizontal.size());
+        for (int i = 0; i < firstVertical.size(); i++) {
+            Set<Node> intersection = new HashSet<Node>(Arrays.asList(graph.getNeighbors(firstVertical.get(i)).toArray()));
+            if (intersection.size() > 0) {
+                try {
+                    intersection.retainAll(firstHorizontal);
+                    for (Node neighbour : intersection) {
+                        firstMatrix.set(i, firstHorizontal.indexOf(neighbour), 1);
+                    }
+                } catch (UnsupportedOperationException ex) {
+                    logger.log(Level.SEVERE, null, ex);
+                }
+            }
+        }
+        // second matrix
+        Matrix secondMatrix = new Matrix(secondVertical.size(), secondHorizontal.size());
+        for (int i = 0; i < secondVertical.size(); i++) {
+            
+            Set<Node> intersection = new HashSet<Node>(Arrays.asList(graph.getNeighbors(secondVertical.get(i)).toArray()));
+            if (intersection.size() > 0) {
+                try {
+                    intersection.retainAll(secondHorizontal);
+                    for (Node neighbour : intersection) {
+                        secondMatrix.set(i, secondHorizontal.indexOf(neighbour), 1);
+                    }
+                } catch (UnsupportedOperationException ex) {
+                    logger.log(Level.SEVERE, null, ex);
+                }
+            }
+        }
+        if (cancelled) {
+            return;
+        }
+        Progress.progress(progressTicket, "Multiplication");
+        
+        Matrix result = firstMatrix.timesParallelIndexed(secondMatrix);
+        if (cancelled) {
+            return;
+        }
+        Progress.progress(progressTicket, "Removing nodes/edges");
+        
+        
+        if (removeNodes) {
+            for (Node n : firstHorizontal) {
+                graph.removeNode(n);
+            }
+        } else {
+            if (removeEdges) {
```

Comment:
No changes required here. Just note that this can cause an exception due to this issue https://github.com/gephi/graphstore/issues/103 that will be fixed in a future update.